File: lib/fluxsink/aufluxsink.cc
#include "lib/core/globals.h"
#include "lib/core/logger.h"
#include "lib/config/flags.h"
#include "lib/data/fluxmap.h"
#include "lib/core/bytes.h"
#include "protocol.h"
#include "lib/fluxsink/fluxsink.h"
#include "lib/data/fluxmapreader.h"
#include "lib/fluxsink/fluxsink.pb.h"
#include "lib/config/proto.h"
#include <fstream>
#include <sys/stat.h>
#include <sys/types.h>

class AuSink : public FluxSink::Sink
{
public:
    AuSink(const std::string& directory, bool indexMarkers):
        _directory(directory),
        _indexMarkers(indexMarkers)
    {
    }

    ~AuSink()
    {
        log("Warning: do not play these files, or you will break your "
            "speakers and/or ears!");
    }

    void addFlux(int track, int head, const Fluxmap& fluxmap) override
    {
        unsigned totalTicks = fluxmap.ticks() + 2;
        unsigned channels = _indexMarkers ? 2 : 1;

        mkdir(_directory.c_str(), 0744);
        std::ofstream of(
            fmt::format("{}/c{:02d}.h{:01d}.au", _directory, track, head),
            std::ios::out | std::ios::binary);
        if (!of.is_open())
            error("cannot open output file");

        /* Write header */

        {
            Bytes header;
            header.resize(24);
            ByteWriter bw(header);

            bw.write_be32(0x2e736e64);
            bw.write_be32(24);
            bw.write_be32(totalTicks * channels);
            bw.write_be32(2); /* 8-bit PCM */
            bw.write_be32(TICK_FREQUENCY);
            bw.write_be32(channels); /* channels */

            of.write((const char*)header.cbegin(), header.size());
        }

        /* Write data */

        {
            Bytes data;
            data.resize(totalTicks * channels);
            memset(data.begin(), 0x80, data.size());

            FluxmapReader fmr(fluxmap);
            unsigned timestamp = 0;
            while (!fmr.eof())
            {
                unsigned ticks;
                int event;
                fmr.getNextEvent(event, ticks);
                if (fmr.eof())
                    break;
                timestamp += ticks;

                if (event & F_BIT_PULSE)
                    data[timestamp * channels + 0] = 0x7f;
                if (_indexMarkers && (event & F_BIT_INDEX))
                    data[timestamp * channels + 1] = 0x7f;
            }

            of.write((const char*)data.cbegin(), data.size());
        }
    }

private:
    std::string _directory;
    bool _indexMarkers;
};

class AuFluxSink : public FluxSink
{
public:
    AuFluxSink(const AuFluxSinkProto& config): _config(config) {}

    std::unique_ptr<Sink> create() override
    {
        return std::make_unique<AuSink>(
            _config.directory(), _config.index_markers());
    }

    std::optional<std::filesystem::path> getPath() const override
    {
        return std::make_optional(_config.directory());
    }

    operator std::string() const override
    {
        return fmt::format("au({})", _config.directory());
    }

private:
    const AuFluxSinkProto& _config;
};

std::unique_ptr<FluxSink> FluxSink::createAuFluxSink(
    const AuFluxSinkProto& config)
{
    return std::unique_ptr<FluxSink>(new AuFluxSink(config));
}

File: java/com/cowlark/fluxengine/fluxsink/AuFluxSink.java
package com.cowlark.fluxengine.fluxsink;

import static com.cowlark.fluxengine.external.FluxEngine.F_BIT_INDEX;
import static com.cowlark.fluxengine.external.FluxEngine.F_BIT_PULSE;
import static com.cowlark.fluxengine.external.FluxEngine.TICK_FREQUENCY;

import com.cowlark.fluxengine.core.Bytes;
import com.cowlark.fluxengine.core.FluxEngineException;
import com.cowlark.fluxengine.core.Logger;
import com.cowlark.fluxengine.data.Fluxmap;
import com.cowlark.fluxengine.data.FluxmapReader;
import com.cowlark.fluxengine.decoders.DecoderProto;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * A flux sink which writes Sun .au audio files, ported from
 * lib/fluxsink/aufluxsink.cc.
 */
public class AuFluxSink extends FluxSink
{
    private final String directory;
    private final boolean indexMarkers;

    public AuFluxSink(String directory, boolean indexMarkers)
    {
        this.directory = directory;
        this.indexMarkers = indexMarkers;
    }

    @Override
    public void addFlux(int track, int head, Fluxmap fluxmap)
    {
        Logger.logf("Warning: do not play these files, or you will break your " +
                "speakers and/or ears!");

        int totalTicks = fluxmap.ticks() + 2;
        int channels = indexMarkers ? 2 : 1;

        try
        {
            Files.createDirectories(Path.of(directory));
        } catch (IOException e)
        {
            throw new FluxEngineException("cannot create directory '" + directory + "'");
        }

        Bytes data = new Bytes(totalTicks * channels);
        for (int i = 0; i < data.size(); i++)
            data.setByte(i, (byte) 0x80);

        FluxmapReader fmr = new FluxmapReader(fluxmap, DecoderProto.getDefaultInstance());
        long timestamp = 0;
        while (!fmr.eof())
        {
            FluxmapReader.Event event = fmr.getNextEvent();
            if (fmr.eof())
                break;
            timestamp += event.ticks();

            if ((event.event() & F_BIT_PULSE) != 0)
                data.setByte((int) timestamp * channels, (byte) 0x7f);
            if (indexMarkers && (event.event() & F_BIT_INDEX) != 0)
                data.setByte((int) timestamp * channels + 1, (byte) 0x7f);
        }

        /* Write header */
        Bytes header = new Bytes(24);
        header.writer()
                .writeBe32(0x2e736e64)
                .writeBe32(24)
                .writeBe32(totalTicks * channels)
                .writeBe32(2) /* 8-bit PCM */.writeBe32(TICK_FREQUENCY)
                .writeBe32(channels); /* channels */

        String filename = String.format("%s/c%02d.h%01d.au", directory, track, head);
        try
        {
            Files.write(Path.of(filename), header.concat(data).toByteArray());
        } catch (IOException e)
        {
            throw new FluxEngineException("cannot open output file");
        }
    }
}
